quarkchain/cluster/miner.py

```diff
@@ -232,23 +240,62 @@ def mine_ethash(
 
     @staticmethod
     def mine_sha3sha3(
-        block,
+        block: Union[MinorBlock, RootBlock],
         input_q: MultiProcessingQueue,
         output_q: MultiProcessingQueue,
         mining_params: Dict,
     ):
-        pass
+        # TODO: control flow similar as `mine_ethash`. should refactor
+
+        def mine(start, end):
+            nonlocal header_hash, target
+            for nonce in range(start, end):
+                nonce_bytes = nonce.to_bytes(8, byteorder="big")
+                h = sha3_256(sha3_256(header_hash + nonce_bytes))
+                if h < target:
+                    return nonce_bytes
+            return None
+
+        # TODO: maybe add rounds to config json
+        rounds = mining_params.get("rounds", 100)
+        # outer loop for mining forever
+        while True:
+            # `None` block means termination
+            if not block:
+                output_q.put(None)
+                return
+
+            header_hash = block.header.get_hash_for_mining()
+            target = (2 ** 256 // (block.header.difficulty or 1) - 1).to_bytes(
+                32, byteorder="big"
+            )
+            start_nonce = 0
+            # inner loop for iterating nonce
+            while True:
+                nonce_found = mine(start_nonce + 1, start_nonce + 1 + rounds)
+                if nonce_found:
+                    block.header.nonce = int.from_bytes(nonce_found, byteorder="big")
+                    Miner._post_process_mined_block(block)
+                    output_q.put(block)
+                    block, _ = input_q.get(block=True)  # blocking
+                    break  # break inner loop to refresh mining params
+                # check if new block arrives. if yes, discard current progress and restart
+                try:
+                    block, _ = input_q.get_nowait()
+                    break  # break inner loop to refresh mining params
+                except Exception:  # queue empty
+                    pass
+                # update param and keep mining
+                start_nonce += rounds
```

Comment:
shall we refresh the block between mining of small batch?

Reply:
line 284 achieves that: if a new block can be retrieved, will refresh to the new block